### DISCUSSION TITLE
Empty result when use trained model DB++

### DISCUSSION THREAD

**Discussion by FahNos** [PaddlePaddle/PaddleOCR#14456] [Q&A]
Hi master !
I am newbie, have just use PaddleOCR
When I use trained mode model in link [https://paddleocr.bj.bcebos.com/dygraph_v2.1/en_det/det_r50_db%2B%2B_icdar15_train.tar](url)
or pretrained model [https://paddleocr.bj.bcebos.com/dygraph_v2.1/en_det/ResNet50_dcn_asf_synthtext_pretrained.pdparams](url)
and command :
`!python3 '/content/PaddleOCR/tools/infer/predict_det.py' \
                                   --use_gpu=False \
                                   --det_model_dir="/content/models/ResNet50_dcn_asf_synthtext_pretrained"  \
                                   --image_dir="/content/2.jpg"`

The result I had is empty. 
![image](https://github.com/user-attachments/assets/0066a9c8-b025-4536-850e-febcac00f447)

but When I use this code: 
`from paddleocr import PaddleOCR,draw_ocr`
`ocr = PaddleOCR( use_angle_cls=False,lang='en')`
`result = ocr.ocr(img_path, cls=False)`
`for line in result:`
  `  print(line)`

the resule is good: 
![image](https://github.com/user-attachments/assets/c038b4c0-37fc-4db4-8f9a-0920e876e5c0)

Can you help me explain this error and how to resolve it ?

Thank you so much

**Comment by GreatV** [comment]
Hi! Let’s analyze the issue and explore why you're getting empty results with your command-line inference using the trained model, while the `PaddleOCR` library approach works correctly.

### Possible Causes and Solutions

1. **Incorrect Model Directory Structure**:
   - The `--det_model_dir` parameter should point to the directory containing the exported model files (`inference.pdmodel`, `inference.pdiparams`, and `inference.pdiparams.info`).
   - Ensure that the directory `/content/models/ResNet50_dcn_asf_synthtext_pretrained` contains these files. If not, you may need to extract the `.tar` file you downloaded or ensure that the correct model is used.

   **Solution**:
   Check the contents of `/content/models/ResNet50_dcn_asf_synthtext_pretrained`. If the files are missing, download and extract the model properly:
   ```bash
   tar -xvf det_r50_db++_icdar15_train.tar -C /content/models
   ```

2. **Mismatch Between Model Type and Code**:
   - The model you downloaded (`det_r50_db++_icdar15_train`) is a detection model. However, from your code, the results might suggest you're trying to perform OCR (text recognition) directly.
   - The `predict_det.py` script is specifically for text detection, not recognition. It will return empty results if the image lacks detectable text regions or if there’s a configuration mismatch.

   **Solution**:
   Use the appropriate script for recognition if your goal is OCR. Combine detection and recognition results using the full PaddleOCR pipeline.

   Alternatively, use:
   ```bash
   python3 /content/PaddleOCR/tools/infer/predict_system.py \
       --use_gpu=False \
       --det_model_dir="/content/models/det_r50_db++_icdar15_train" \
       --image_dir="/content/2.jpg"
   ```
   The `predict_system.py` script handles both detection and recognition seamlessly.

3. **Incorrect Preprocessing or Parameters**:
   - When using `predict_det.py`, ensure the preprocessing steps match the model's training setup. The `det_db_thresh`, `det_db_box_thresh`, and other parameters may need adjustment.
   - By default, the thresholds might not be optimal for your specific image.

   **Solution**:
   Try tuning the detection thresholds:
   ```bash
   python3 /content/PaddleOCR/tools/infer/predict_det.py \
       --use_gpu=False \
       --det_model_dir="/content/models/det_r50_db++_icdar15_train" \
       --image_dir="/content/2.jpg" \
       --det_db_thresh=0.3 \
       --det_db_box_thresh=0.5
   ```

4. **Image Format or Path Issue**:
   - Ensure that `/content/2.jpg` is a valid image file and in a format supported by the detection model.
   - If the file path or format is incorrect, the model will not process the image, resulting in empty output.

   **Solution**:
   Validate the image:
   ```python
   from PIL import Image
   img = Image.open("/content/2.jpg")
   img.show()
   ```
   Ensure the path is correct and the image is suitable for text detection.

5. **Differences Between PaddleOCR API and CLI Tools**:
   - The `PaddleOCR` Python API (`ocr = PaddleOCR(...)`) internally handles both text detection and recognition, along with preprocessing. This is why it works well with minimal configuration.
   - However, the CLI tools (`predict_det.py`) require explicit setup for the specific task (detection or recognition).

   **Solution**:
   If the `PaddleOCR` API works better for your use case, consider using it directly in your Python script instead of the CLI tools.

---

### Summary of Steps to Resolve:
1. **Verify Model Directory**:
   Ensure `/content/models/ResNet50_dcn_asf_synthtext_pretrained` contains `inference.pdmodel`, `inference.pdiparams`, and `inference.pdiparams.info`.

2. **Use the Full Pipeline**:
   Use `predict_system.py` if you want both detection and recognition:
   ```bash
   python3 /content/PaddleOCR/tools/infer/predict_system.py \
       --use_gpu=False \
       --det_model_dir="/content/models/det_r50_db++_icdar15_train" \
       --image_dir="/content/2.jpg"
   ```

3. **Tune Parameters**:
   Adjust detection thresholds (`det_db_thresh`, `det_db_box_thresh`) if detection results are still empty.

4. **Validate the Image**:
   Ensure the image path and format are correct.

By applying these fixes, you should be able to resolve the issue and obtain results similar to those from the `PaddleOCR` library. Let me know if you need further assistance!

---
*Response generated by [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Comment by FahNos** [comment]
I use it command:
`!python3 '/content/PaddleOCR/tools/infer/predict_system.py' \
                --use_gpu=True \
                --det_algorithm="SAST" \
                --det_model_dir="/content/models/det_r50_vd_sast_icdar15_v2.0_train" \
                --rec_algorithm="SRN" \
                --rec_model_dir="/content/inference/rec_srn_vnese" \
                --rec_image_shape="1, 64, 256" \
                --image_dir="/content/inference/inference_result/1.jpg" \
                --rec_char_type="ch" \
                --drop_score=0.7 \
                --rec_char_dict_path="/content/vi_vietnam.txt"`

But have error: 
`predict_system.py: error: unrecognized arguments: --rec_char_type=ch`

Can you help me resolve it ?
